### PR TITLE
RFC: SDPA Optimization CPU

### PR DESCRIPTION
**Pull request by Valentine233**

### CODE REVIEW

**Comment by Valentine233** [comment]
@mingfeima @jgong5, please have a look and help review, thanks.

File: RFC-0025-sdpa-optm-cpu.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+Here are the detailed implementation items:
+
+*   The flash attention CPU kernel is added, in which both forward and backward paths are implemented for data types float32 and bfloat16.
+*   Write an SDPA selecting function for CPU to automatically choose one SDPA implementation among several ones.
+
+The following will be nice to have for PT 2.1:
+
+*   Support data type of float16.
+*   Enable the SDPA graph rewriting for Inductor.
```
[the remark this answers is not in the document]
Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md:19]: > I suppose the fusion would automatically apply to the inductor after we add the kernel support to SDPA, right?

Yes, the fusion is automatically applied now. However, there would be assertion checks for intermediate outputs' shapes and strides, which are not the same for CPU and CUDA implementations. For example, the output `logsumexp` which is used for backward calculation, has the shape of `{batch_size, query_length, num_head}` in CPU and the shape of `{batch_size, num_head, query_length}` in CUDA. Besides, other outputs like `cum_seq_q`, `cum_seq_k`,  are not used and are assigned with a zero-size tensor in CPU. For now, I have temporarily disabled CPU fusion path and would enable it if having time to fix the issue mentioned before.

File: RFC-0025-sdpa-optm-cpu.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+*   Further block-related tuning for the fused SDPA.
+
+
+## **Performance**
+All validations are run on SPR machine.
+
+### NanoGPT's SDPA kernel
```
[the remark this answers is not in the document]
Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md:27]: Thanks, added.

File: RFC-0025-sdpa-optm-cpu.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+Some additional works:
+*   [#104583 enable mkl’s bfloat16 gemm on PT](https://github.com/pytorch/pytorch/pull/104583)
+*   [#104584 expand functional utils on CPU vectorization path](https://github.com/pytorch/pytorch/pull/104584)
+
+## **Discussion**
+
+For the SDPA optimization, there are two things that needed to be discussed and I hope to have your precious opinions.
+
+One is about the util functions for SDPA selection. The current util functions are under the CUDA folder, i.e. `transformers/cuda/sdp_utils`. For CPU, we have similar functions in `transformers/sdp_utils_cpp` (see #105131). It is good to know whether we need to make them a unified API.
```
[the remark this answers is not in the document]
Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md:77]: There exist compilation issues when the files in CUDA folder include the files in non-CUDA folder, and vice versa. I'll spend more time to figure out the solution.

File: RFC-0025-sdpa-optm-cpu.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+## **Motivation**
+As LLM tends to accept a large batch size and a long context length, the requirement of large memory may lead to OOM issues or result in bad performance. To reduce memory usage and provide a substantial speedup for attention-related models, it is important to optimize SDPA. The fused SDPA, e.g. flash attention, is one type of the optimized SDPA algorithms designed for memory-bound problems, with better parallelism and memory access patterns. In PT 2.0, there exist both the basic unfused SDPA and the fused SDPA for CUDA, while only the unfused SDPA has CPU implementation. To fill the gap between CPU and CUDA, it is proposed to optimize SDPA by implementing the fused SDPA for CPU in PT 2.1.
+
+## **Implementation**
+We submitted PRs for CPU SDPA optimization and demonstrated up to 3x performance speedup on attention-related benchmarks.
+
+Here are the detailed implementation items:
+
+*   The flash attention CPU kernel is added, in which both forward and backward paths are implemented for data types float32 and bfloat16.
+*   Write an SDPA selecting function for CPU to automatically choose one SDPA implementation among several ones.
```
[the remark this answers is not in the document]
Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md]: Thanks, modified.

File: RFC-0025-sdpa-optm-cpu.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+## **Motivation**
+As LLM tends to accept a large batch size and a long context length, the requirement of large memory may lead to OOM issues or result in bad performance. To reduce memory usage and provide a substantial speedup for attention-related models, it is important to optimize SDPA. The fused SDPA, e.g. flash attention, is one type of the optimized SDPA algorithms designed for memory-bound problems, with better parallelism and memory access patterns. In PT 2.0, there exist both the basic unfused SDPA and the fused SDPA for CUDA, while only the unfused SDPA has CPU implementation. To fill the gap between CPU and CUDA, it is proposed to optimize SDPA by implementing the fused SDPA for CPU in PT 2.1.
+
+## **Implementation**
+We submitted PRs for CPU SDPA optimization and demonstrated up to 3x performance speedup on attention-related benchmarks.
+
+Here are the detailed implementation items:
+
+*   The flash attention CPU kernel is added, in which both forward and backward paths are implemented for data types float32 and bfloat16.
```
[the remark this answers is not in the document]
Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md]: Thanks, modified.

**Comment by albanD** [comment]
cc @drisspg

**Review by jgong5 — APPROVED** [review]
LGTM. I know we are also working on EfficientAttention. Do we plan to submit a separate RFC or combine them two? @mingfeima

**Comment by mingfeima** [comment]
> LGTM. I know we are also working on EfficientAttention. Do we plan to submit a separate RFC or combine them two? @mingfeima

No, just this one.

File: RFC-0025-sdpa-optm-cpu.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (drisspg) [reply RFC-0025-sdpa-optm-cpu.md:77]: I think that this can be resolved, there is functionality that is needed for cuda dispatching that requires cuda utils to be built. I think though that the common utils can be abstracted out and shared between the two

**Comment by drisspg** [comment]
> jgong5

Can you elaborate on the difference between EfficientAttention and what is discussed here?

File: RFC-0025-sdpa-optm-cpu.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+## **Implementation**
+We submitted PRs for CPU SDPA optimization and demonstrated up to 3x performance speedup on attention-related benchmarks.
+
+Here are the detailed implementation items:
+
+*   The flash attention CPU kernel is added, in which both forward and backward paths are implemented for data types float32 and bfloat16. Blocking is applied on dimensions of query length and kv length and the fusion of gemm + softmax update + gemm is done at once for each block. Specifically, FP32In-FP32Out and FP32In-BF16Out adopt the mkl gemm and BF16In-BF16Out adopts the OneDNN one. Parallelization is on the dimensions of batch size, head number and query length for forward path, and on the dimensions of batch size and head number for backward path. In addition, the causal attention mask is supported. As the attention is masked for the unseen tokens, early termination is applied and we only calculate the blocks in the lower triangular part.
```
Comment (drisspg) [inline RFC-0025-sdpa-optm-cpu.md:13]: How is the mixed_dtype supposed to be handled ? Will it be built off of this: https://github.com/pytorch/pytorch/pull/103333?

File: RFC-0025-sdpa-optm-cpu.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md:19]: Yeah, after doing the `fused attention` pattern matching in Inductor, some graph nodes could be replaced with one SDPA node. If the SDPA node uses sdpa_flash, Inductor would check the all outputs' shapes and strides in CPP output code.

File: RFC-0025-sdpa-optm-cpu.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md:77]: Thanks, would do it.

File: RFC-0025-sdpa-optm-cpu.md (hunk 6 above)
Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md:13]: Currently, we only support the case where the three inputs, e.g. q, k, v, and the output own the same data type. Do you have any suggestions?

Comment (drisspg) [reply RFC-0025-sdpa-optm-cpu.md:13]: This makes sense, I guess I was confused on `FP32In-BF16Out` but is this used for the second bmm after the softmax?

Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md]: Sorry, I accidently made a typo. Fixed from `FP32In-BF16Out` to `BF16In-FP32Out`, which is used for the two bmms in forward path when input dtype is BF16.

File: RFC-0025-sdpa-optm-cpu.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Valentine233) [reply RFC-0025-sdpa-optm-cpu.md:77]: The common utils abstraction has been done in [#105131](https://github.com/pytorch/pytorch/pull/105131).

**Comment by Valentine233** [comment]
> Yeah I think this will be great to have, also still curious if there is a difference between EfficientAttention mentioned above but otherwise this looks great

Efficient Attention mentioned above is another fused SDPA algorithm, which is enabled in CUDA but not in CPU now. After the Flash attention being merged, we would also upstream Efficient Attention for CPU soon.

**Comment by Valentine233** [comment]
@drisspg BTW, is this RFC needed to be merged? If so, could you please kindly help merge it as we don't have the permissions.

**Comment by drisspg** [comment]
> > Yeah I think this will be great to have, also still curious if there is a difference between EfficientAttention mentioned above but otherwise this looks great
> 
> Efficient Attention mentioned above is another fused SDPA algorithm, which is enabled in CUDA but not in CPU now. After the Flash attention being merged, we would also upstream Efficient Attention for CPU soon.

If you are referring to mem_eff_attention this is just another implementation ( by xformers) of the original flash attention algorithm. This was also included because it had different performance characteristics and different hardware support. I would be surprised if for CPU both are needed.

**Comment by mingfeima** [comment]
@drisspg Indeed, the mem_eff_attention is another form of flash attention; we have done both on the CPU side but mostly from the aspect that we should **fill the feature gap between CPU and CUDA device on PyTorch**. As @Valentine233 commented, we still prioritize flash attention and will continue to work on it for new features  such as GQA, etc.

**Comment by drisspg** [comment]
Okay I would say you don't need to create a CPU impl for mem_eff_attention if it is going to behave identical to how sdpa_flash_attention's cpu impl. We just update the sdp_dispatcher to never dispatch to mem_eff for cpu. This is a small point but just want to make sure there isn't any wasted work.

All that being said I will merge this RFC in! Thanks for all the discussion

**Comment by mingfeima** [comment]
> Okay I would say you don't need to create a CPU impl for mem_eff_attention if it is going to behave identical to how sdpa_flash_attention's cpu impl. We just update the sdp_dispatcher to never dispatch to mem_eff for cpu. This is a small point but just want to make sure there isn't any wasted work.
> 
> All that being said I will merge this RFC in! Thanks for all the discussion

Wow, cool. Did not know this before, then it totally makes sense to drop cpu mem_eff_attention :)

**Comment by jgong5** [comment]
> Can you elaborate on the difference between EfficientAttention and what is discussed here?

Just back from PTO. Thanks for your comments. That's good we only need to keep one implementation that is best for CPU.
